File: packages/protocol/src/graph.ts
import type { ChangeArtifactInput } from "./artifact.js";
import { FormatError } from "./cbor.js";
import { canReference, type Realm } from "./realm.js";

export type GraphArtifactKind = "tree" | "change";

export interface GraphArtifactSummary {
  project: string;
  realm: Realm;
  kind: GraphArtifactKind;
  actorKey: Uint8Array;
  logicalClock: bigint;
}

export type GraphResolver = (
  artifactId: string,
) => GraphArtifactSummary | undefined;

function sameBytes(left: Uint8Array, right: Uint8Array): boolean {
  return (
    left.length === right.length &&
    left.every((byte, index) => byte === right[index])
  );
}

function graphError(
  code:
    | "cross_project_reference"
    | "parent_realm_mismatch"
    | "realm_flow_denied"
    | "unknown_required_semantics"
    | "invalid_logical_clock",
  message: string,
): never {
  throw new FormatError(code, message);
}

/** Resolves and validates the graph edges of one schema-0 change. */
export function validateChangeGraph(
  change: ChangeArtifactInput,
  resolve: GraphResolver,
): void {
  const root = resolve(change.root);
  if (root === undefined || root.kind !== "tree") {
    graphError(
      "unknown_required_semantics",
      "root is unavailable or is not a tree",
    );
  }
  if (root.project !== change.project) {
    graphError("cross_project_reference", "root belongs to another project");
  }
  if (!canReference(change.realm, root.realm, "content")) {
    graphError(
      "realm_flow_denied",
      "root realm is not readable from change realm",
    );
  }

  for (const parentId of change.parents) {
    const parent = resolve(parentId);
    if (parent === undefined || parent.kind !== "change") {
      graphError(
        "unknown_required_semantics",
        "parent is unavailable or is not a change",
      );
    }
    if (parent.project !== change.project) {
      graphError(
        "cross_project_reference",
        "parent belongs to another project",
      );
    }
    if (!canReference(change.realm, parent.realm, "parent")) {
      graphError(
        "parent_realm_mismatch",
        "parent realm differs from change realm",
      );
    }
    if (
      sameBytes(parent.actorKey, change.actorKey) &&
      change.logicalClock <= parent.logicalClock
    ) {
      graphError(
        "invalid_logical_clock",
        "logical clock must advance same-actor parents",
      );
    }
  }
}

File: crates/ef-format/src/graph.rs
//! Resolved graph validation for schema-0 changes.

use crate::{ChangeArtifact, FormatError, FormatErrorCode, Realm, ReferenceClass, can_reference};

/// Artifact kinds needed while resolving a change graph.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum GraphArtifactKind {
    Tree,
    Change,
}

/// Resolved metadata needed to validate an immutable graph edge.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct GraphArtifactSummary {
    pub project: String,
    pub realm: Realm,
    pub kind: GraphArtifactKind,
    pub actor_key: [u8; 32],
    pub logical_clock: u64,
}

/// Resolves and validates the graph edges of one schema-0 change.
///
/// # Errors
///
/// Returns a stable graph error when an edge cannot be resolved or violates
/// project, kind, realm-flow, or same-actor logical-clock rules.
pub fn validate_change_graph(
    change: &ChangeArtifact,
    resolve: impl Fn(&str) -> Option<GraphArtifactSummary>,
) -> Result<(), FormatError> {
    let root = resolve(&change.root).ok_or_else(|| {
        FormatError::new(
            FormatErrorCode::UnknownRequiredSemantics,
            "root is unavailable or is not a tree",
        )
    })?;
    if root.kind != GraphArtifactKind::Tree {
        return Err(FormatError::new(
            FormatErrorCode::UnknownRequiredSemantics,
            "root is unavailable or is not a tree",
        ));
    }
    if root.project != change.meta.project {
        return Err(FormatError::new(
            FormatErrorCode::CrossProjectReference,
            "root belongs to another project",
        ));
    }
    if !can_reference(change.meta.realm, root.realm, ReferenceClass::Content) {
        return Err(FormatError::new(
            FormatErrorCode::RealmFlowDenied,
            "root realm is not readable from change realm",
        ));
    }

    for parent_id in &change.meta.parents {
        let parent = resolve(parent_id).ok_or_else(|| {
            FormatError::new(
                FormatErrorCode::UnknownRequiredSemantics,
                "parent is unavailable or is not a change",
            )
        })?;
        if parent.kind != GraphArtifactKind::Change {
            return Err(FormatError::new(
                FormatErrorCode::UnknownRequiredSemantics,
                "parent is unavailable or is not a change",
            ));
        }
        if parent.project != change.meta.project {
            return Err(FormatError::new(
                FormatErrorCode::CrossProjectReference,
                "parent belongs to another project",
            ));
        }
        if !can_reference(change.meta.realm, parent.realm, ReferenceClass::Parent) {
            return Err(FormatError::new(
                FormatErrorCode::ParentRealmMismatch,
                "parent realm differs from change realm",
            ));
        }
        if parent.actor_key == change.meta.actor_key
            && change.meta.logical_clock <= parent.logical_clock
        {
            return Err(FormatError::new(
                FormatErrorCode::InvalidLogicalClock,
                "logical clock must advance same-actor parents",
            ));
        }
    }
    Ok(())
}
